zap/src/output/luau/mod.rs

```diff
@@ -18,69 +15,526 @@ pub trait Output {
-			Stmt::Local(name, expr) => {
-				if let Some(expr) = expr {
-					self.push_line(&format!("local {name} = {expr}"));
+					self.push_line(&format!("alloc({exact})"));
+					self.push_line(&format!(
+						"buffer.copy(outgoing_buff, outgoing_apos, {from}, 0, {exact})"
+					));
 				} else {
-					self.push_line(&format!("local {name}"));
+					self.push_line(&format!("local len = #{from}"));
+
+					if checks {
+						self.push_range_check(*len, "len");
+					}
+
+					self.push_line("alloc(len)");
+					self.push_line(&format!("buffer.copy(outgoing_buff, outgoing_apos, {from}, 0, len)"));
 				}
 			}
-			Stmt::LocalTuple(var, expr) => {
-				let items = var.join(", ");
 
-				if let Some(expr) = expr {
-					self.push_line(&format!("local {items} = {expr}"));
+			Ty::Arr(ty, len) => {
+				if let Some(exact) = len.exact() {
+					if checks {
+						self.push_line(&format!("assert(#{from} == {exact})"));
+					}
+
+					self.push_line_indent(&format!("for i = 1, {exact} do"));
+					self.push_line(&format!("local value = {from}[i]"));
+					self.push_ser("value", ty, checks);
+					self.push_dedent_line("end");
 				} else {
-					self.push_line(&format!("local {items}"));
+					self.push_line(&format!("local len = #{from}"));
+
+					if checks {
+						self.push_range_check(*len, "len");
+					}
+
+					self.push_line("alloc(len)");
+					self.push_line("buffer.writeu16(outgoing_buff, outgoing_apos, len)");
+					self.push_line_indent("for i = 1, len do");
+					self.push_line(&format!("local value = {from}[i]"));
+					self.push_ser("value", ty, checks);
+					self.push_dedent_line("end");
 				}
 			}
 
-			Stmt::Assign(var, expr) => self.push_line(&format!("{var} = {expr}")),
-			Stmt::Error(msg) => self.push_line(&format!("error(\"{msg}\")")),
-			Stmt::Assert(cond, msg) => match msg {
-				Some(msg) => self.push_line(&format!("assert({cond}, \"{msg}\")")),
-				None => self.push_line(&format!("assert({cond})")),
-			},
+			Ty::Map(key, val) => {
+				self.push_line("local len = 0");
+				self.push_line("local len_pos = alloc(2)");
+
+				self.push_line_indent(&format!("for k, v in {from} do"));
+				self.push_line("len += 1");
+				self.push_ser("k", key, checks);
+				self.push_ser("v", val, checks);
+				self.push_dedent_line("end");
+
+				self.push_line("buffer.writeu16(outgoing_buff, len_pos, len)");
+			}
+
+			Ty::Opt(ty) => {
+				self.push_line("alloc(1)");
+				self.push_line_indent(&format!("if {from} == nil then"));
+				self.push_line("buffer.writeu8(outgoing_buff, outgoing_apos, 0)");
+				self.push_dedent_line_indent("else");
+				self.push_line("buffer.writeu8(outgoing_buff, outgoing_apos, 1)");
+				self.push_ser(from, ty, checks);
+				self.push_dedent_line("end");
+			}
+
+			Ty::Ref(name) => {
+				self.push_line(&format!("types.write_{name}({from})"));
+			}
+
+			Ty::Enum(enum_ty) => match enum_ty {
+				Enum::Unit(enumerators) => {
+					let numty = NumTy::from_f64(0.0, enumerators.len() as f64 - 1.0);
```

Comment:
Just check - rust is 0-indexed so you don't need to `- 1.0`?

Reply:
because rust is 0-indexed we need to subtract 1, the length of a vec is always 1 greater than the highest index